File: cli/src/wsc/ops/fetch.py
"""wsc fetch — clean a known URL via Firecrawl, falling back to a stdlib
``urllib`` + ``html.parser`` extractor (degraded; no JS, no PDF).
"""

from __future__ import annotations

import html as _html
import re
import urllib.error
import urllib.request
from typing import Any, Dict, Optional

from wsc import audit
from wsc._url import normalize_url
from wsc.ops._chain import filtered_chain, query_fingerprint, run_chain
from wsc.providers import FetchedPage


def run(
    url: str,
    *,
    formats: Optional[list] = None,
    screenshot: bool = False,
    correlation_id: Optional[str] = None,
    no_receipt: bool = False,
) -> Dict[str, Any]:
    chain = filtered_chain("url_fetch") or ["firecrawl"]

    def firecrawl_action(provider):
        return provider.scrape(url, formats=formats, screenshot=screenshot)

    actions = {"firecrawl": firecrawl_action}

    with audit.start_call(
        "fetch", provider="firecrawl",
        correlation_id=correlation_id, no_receipt=no_receipt,
    ) as receipt:
        receipt.update(query_fingerprint(url))
        receipt["params"] = {"formats": formats, "screenshot": screenshot}
        active, page, fallback = run_chain(chain, actions)
        receipt["fallback_chain"] = list(fallback)

        if active is None:
            # Stdlib fallback. Degraded on purpose: no JS, no PDF, no CSS removal.
            try:
                page = _stdlib_fetch(url)
                receipt["fallback_chain"].append({"from": "firecrawl", "to": "urllib", "reason": "all_providers_failed"})
                receipt["provider"] = "urllib"
                receipt["status"] = "degraded"
                receipt["selected_urls"] = [page.url]
                receipt["selected_count"] = 1
                receipt["results_count"] = 1
                return {
                    "ok": True,
                    "operation": "fetch",
                    "provider": "urllib",
                    "page": page.to_dict(),
                    "fallback_chain": receipt["fallback_chain"],
                    "status": "degraded",
                    "returncode": 0,
                }
            except Exception as exc:  # noqa: BLE001
                receipt["status"] = "error"
                receipt["error"] = str(exc)[:200]
                return {
                    "ok": False,
                    "operation": "fetch",
                    "provider": None,
                    "fallback_chain": receipt["fallback_chain"]
                    + [{"from": "urllib", "reason": "transport_error", "error": str(exc)[:200]}],
                    "error": f"fetch failed: {exc}",
                    "returncode": 2,
                }

        receipt["provider"] = active
        receipt["selected_urls"] = [page.url] if page.url else []
        receipt["selected_count"] = 1
        receipt["results_count"] = 1
        return {
            "ok": True,
            "operation": "fetch",
            "provider": active,
            "page": page.to_dict(),
            "fallback_chain": receipt["fallback_chain"],
            "status": page.status,
            "returncode": 0,
        }


# --- Stdlib fallback: poor-man's URL → markdown-ish text ----------------


_TAG_RE = re.compile(r"<[^>]+>")
_WS_RE = re.compile(r"\n{3,}")


def _stdlib_fetch(url: str, *, timeout: float = 30.0) -> FetchedPage:
    req = urllib.request.Request(url, headers={"user-agent": "wsc/0.1 (+stdlib fallback)"})
    with urllib.request.urlopen(req, timeout=timeout) as resp:
        body = resp.read()
        encoding = resp.headers.get_content_charset() or "utf-8"
        text = body.decode(encoding, errors="replace")
    title_match = re.search(r"<title[^>]*>(.*?)</title>", text, re.IGNORECASE | re.DOTALL)
    title = _html.unescape(title_match.group(1).strip()) if title_match else url
    body_text = _html.unescape(_TAG_RE.sub("\n", text)).strip()
    body_text = _WS_RE.sub("\n\n", body_text)
    return FetchedPage(
        url=url,
        title=title,
        markdown=body_text[:50000],  # safety clamp
        provider="urllib",
        status="degraded",
        url_normalized=normalize_url(url),
    )

File: src/ops/fetch.ts
/** wsc fetch — Firecrawl primary, native fetch+regex stdlib fallback (degraded). */

import { withCall } from "../audit.js";
import * as cache from "../cache.js";
import { filteredChain, queryFingerprint, runChain } from "./_chain.js";
import type { Action, FallbackStep } from "./_chain.js";
import { FetchedPage, FirecrawlProvider, httpRequest } from "../providers/index.js";
import { normalizeUrl } from "../_url.js";

export interface FetchOptions {
  formats?: string[];
  screenshot?: boolean;
  correlationId?: string;
  noReceipt?: boolean;
  noCache?: boolean;
}

interface FetchCacheValue {
  provider: string;
  page: ReturnType<FetchedPage["toJSON"]>;
  fallback_chain: FallbackStep[];
  cached_status: "ok" | "degraded";
}

const TAG_RE = /<[^>]+>/g;
const WS_RE = /\n{3,}/g;

const ENTITY_MAP: Record<string, string> = { "&amp;": "&", "&lt;": "<", "&gt;": ">", "&quot;": '"', "&#39;": "'", "&nbsp;": " " };

function decodeEntities(s: string): string {
  return s
    .replace(/&(amp|lt|gt|quot|#39|nbsp);/g, (m) => ENTITY_MAP[m] ?? m)
    .replace(/&#(\d+);/g, (_, n) => String.fromCodePoint(Number(n)))
    .replace(/&#x([0-9a-f]+);/gi, (_, n) => String.fromCodePoint(Number.parseInt(n, 16)));
}

async function stdlibFetch(url: string): Promise<FetchedPage> {
  const resp = await httpRequest(url, { method: "GET", timeoutMs: 30_000 });
  const text = resp.text;
  const titleMatch = text.match(/<title[^>]*>([\s\S]*?)<\/title>/i);
  const title = titleMatch ? decodeEntities(titleMatch[1]!.trim()) : url;
  const body = decodeEntities(text.replace(TAG_RE, "\n")).trim().replace(WS_RE, "\n\n");
  return new FetchedPage({
    url,
    title,
    markdown: body.slice(0, 50_000),
    provider: "urllib",
    status: "degraded",
  });
}

export async function run(url: string, opts: FetchOptions = {}): Promise<Record<string, unknown>> {
  const chain = filteredChain("url_fetch");
  const cKey = cache.cacheKey({
    op: "fetch",
    query: url,
    params: { formats: opts.formats ?? null, screenshot: !!opts.screenshot },
  });
  const firecrawlAction: Action<FetchedPage> = (provider) =>
    (provider as FirecrawlProvider).scrape(url, { formats: opts.formats, screenshot: opts.screenshot });

  return await withCall(
    "fetch",
    { provider: "firecrawl", correlationId: opts.correlationId, noReceipt: opts.noReceipt },
    async (receipt) => {
      Object.assign(receipt, queryFingerprint(url));
      receipt.params = { formats: opts.formats ?? null, screenshot: !!opts.screenshot };

      const cached = await cache.get<FetchCacheValue>(cKey, cache.CACHE_TTL_SEC.fetch!, { noCache: opts.noCache });
      if (cached) {
        receipt.cache_hit = true;
        receipt.provider = cached.provider;
        receipt.fallback_chain = cached.fallback_chain;
        const cachedUrl = (cached.page as Record<string, unknown>).url;
        receipt.selected_urls = typeof cachedUrl === "string" ? [cachedUrl] : [];
        receipt.selected_count = receipt.selected_urls.length;
        receipt.results_count = receipt.selected_urls.length;
        if (cached.cached_status === "degraded") receipt.status = "degraded";
        return {
          ok: true,
          operation: "fetch",
          provider: cached.provider,
          page: cached.page,
          fallback_chain: cached.fallback_chain,
          status: cached.cached_status,
          cache_hit: true,
          returncode: 0,
        };
      }

      const { active, result, fallback } = await runChain(chain, { firecrawl: firecrawlAction });
      receipt.fallback_chain = [...fallback];
      receipt.cache_hit = false;

      if (active === null || result === null) {
        try {
          const page = await stdlibFetch(url);
          (receipt.fallback_chain as FallbackStep[]).push({
            from: "firecrawl",
            to: "urllib",
            reason: "all_providers_failed",
          });
          receipt.provider = "urllib";
          receipt.status = "degraded";
          receipt.selected_urls = [page.url];
          receipt.selected_count = 1;
          receipt.results_count = 1;
          const pageJson = page.toJSON();
          await cache.set<FetchCacheValue>(
            cKey,
            { provider: "urllib", page: pageJson, fallback_chain: receipt.fallback_chain as FallbackStep[], cached_status: "degraded" },
            { ttlSec: cache.CACHE_TTL_SEC.fetch!, op: "fetch", provider: "urllib", noCache: opts.noCache },
          );
          return {
            ok: true,
            operation: "fetch",
            provider: "urllib",
            page: pageJson,
            fallback_chain: receipt.fallback_chain,
            status: "degraded",
            cache_hit: false,
            returncode: 0,
          };
        } catch (err) {
          receipt.status = "error";
          receipt.error = (err instanceof Error ? err.message : String(err)).slice(0, 200);
          (receipt.fallback_chain as FallbackStep[]).push({
            from: "urllib",
            reason: "transport_error",
            error: (err instanceof Error ? err.message : String(err)).slice(0, 200),
          });
          return {
            ok: false,
            operation: "fetch",
            provider: null,
            fallback_chain: receipt.fallback_chain,
            error: `fetch failed: ${err instanceof Error ? err.message : String(err)}`,
            returncode: 2,
          };
        }
      }

      receipt.provider = active;
      receipt.selected_urls = result.url ? [result.url] : [];
      receipt.selected_count = 1;
      receipt.results_count = 1;
      const pageJson = result.toJSON();
      const status = result.status;
      await cache.set<FetchCacheValue>(
        cKey,
        { provider: active, page: pageJson, fallback_chain: fallback, cached_status: status === "degraded" ? "degraded" : "ok" },
        { ttlSec: cache.CACHE_TTL_SEC.fetch!, op: "fetch", provider: active, noCache: opts.noCache },
      );
      return {
        ok: true,
        operation: "fetch",
        provider: active,
        page: pageJson,
        fallback_chain: receipt.fallback_chain,
        status: result.status,
        cache_hit: false,
        returncode: 0,
      };
    },
  );
}

export function normalizeForReceipt(url: string): string {
  return normalizeUrl(url);
}
